### PR TITLE
refactor: use patcher for pod controller

### PR DESCRIPTION
**Pull request by FabianKramm**
**What issue type does this pull request address?** (keep at least one, remove the others) 
/kind enhancement

Part of ENG-4168

Changes:
* Pod Controller
* Persistent Volume Claim Controller
* Volume Snapshots Controller

### CODE REVIEW

File: pkg/controllers/resources/persistentvolumeclaims/syncer_test.go [hunk 1]
```diff
@@ -95,15 +95,13 @@ func TestSync(t *testing.T) {
 			Name:      pObjectMeta.Name,
 			Namespace: pObjectMeta.Namespace,
 			Annotations: map[string]string{
-				translate.NameAnnotation:               vObjectMeta.Name,
-				translate.NamespaceAnnotation:          vObjectMeta.Namespace,
-				translate.UIDAnnotation:                "",
-				translate.KindAnnotation:               corev1.SchemeGroupVersion.WithKind("PersistentVolumeClaim").String(),
-				translate.ManagedAnnotationsAnnotation: "otherAnnotationKey",
```
Comment (facchettos) [inline pkg/controllers/resources/persistentvolumeclaims/syncer_test.go:102]: why is this removed?

File: pkg/controllers/resources/persistentvolumeclaims/syncer_test.go [hunk 2]
```diff
@@ -95,15 +95,13 @@ func TestSync(t *testing.T) {
...
-				bindCompletedAnnotation:                "testannotation",
-				boundByControllerAnnotation:            "testannotation2",
-				storageProvisionerAnnotation:           "testannotation3",
-				"otherAnnotationKey":                   "don't update this",
```
Comment (facchettos) [inline pkg/controllers/resources/persistentvolumeclaims/syncer_test.go:106]: and this too

File: pkg/controllers/resources/persistentvolumeclaims/translate.go [hunk 3]
```diff
@@ -99,61 +98,38 @@ func (s *persistentVolumeClaimSyncer) translateSelector(ctx *synccontext.SyncCon
...
 		}
-		updated.Spec.Resources.Requests["storage"] = vObj.Spec.Resources.Requests["storage"]
-	}
-
-	changed, updatedAnnotations, updatedLabels := s.TranslateMetadataUpdate(ctx, vObj, pObj)
-	if changed {
-		updated = translator.NewIfNil(updated, pObj)
-		updated.Annotations = updatedAnnotations
-		updated.Labels = updatedLabels
+		pObj.Spec.Resources.Requests["storage"] = vObj.Spec.Resources.Requests["storage"]
```
Comment (facchettos) [inline pkg/controllers/resources/persistentvolumeclaims/translate.go:107]: couldn't this be 
```go
pObj.Spec.Resources.Requests = vObj.Spec.Resources.Requests
```
?

File: pkg/controllers/resources/persistentvolumeclaims/translate.go [hunk 4]
```diff
@@ -99,61 +98,38 @@ func (s *persistentVolumeClaimSyncer) translateSelector(ctx *synccontext.SyncCon
...
-		if updated.Annotations[bindCompletedAnnotation] != pObj.Annotations[bindCompletedAnnotation] {
-			updated.Annotations[bindCompletedAnnotation] = pObj.Annotations[bindCompletedAnnotation]
-		}
-		if updated.Annotations[boundByControllerAnnotation] != pObj.Annotations[boundByControllerAnnotation] {
-			updated.Annotations[boundByControllerAnnotation] = pObj.Annotations[boundByControllerAnnotation]
-		}
-		if updated.Annotations[storageProvisionerAnnotation] != pObj.Annotations[storageProvisionerAnnotation] {
-			updated.Annotations[storageProvisionerAnnotation] = pObj.Annotations[storageProvisionerAnnotation]
+func (s *persistentVolumeClaimSyncer) translateUpdateBackwards(pObj, vObj *corev1.PersistentVolumeClaim) {
+	if vObj.Annotations[bindCompletedAnnotation] != pObj.Annotations[bindCompletedAnnotation] {
```
Comment (facchettos) [inline pkg/controllers/resources/persistentvolumeclaims/translate.go:117]: don't think this is needed

File: pkg/controllers/resources/persistentvolumeclaims/translate.go [hunk 5]
```diff
@@ -99,61 +98,38 @@ func (s *persistentVolumeClaimSyncer) translateSelector(ctx *synccontext.SyncCon
...
+		vObj.Annotations[bindCompletedAnnotation] = pObj.Annotations[bindCompletedAnnotation]
 	}
-
-	return updated
-}
-
-func translateUpdateNeeded(pAnnotations, vAnnotations map[string]string) bool {
-	if pAnnotations == nil {
-		pAnnotations = map[string]string{}
+	if vObj.Annotations[boundByControllerAnnotation] != pObj.Annotations[boundByControllerAnnotation] {
```
Comment (facchettos) [inline pkg/controllers/resources/persistentvolumeclaims/translate.go:123]: all these checks aren't needed I guess

File: pkg/controllers/resources/persistentvolumeclaims/syncer_test.go (hunk 1 above)
Comment (FabianKramm) [reply pkg/controllers/resources/persistentvolumeclaims/syncer_test.go:102]: The "otherAnnotationKey" and "translate.ManagedAnnotationsAnnotation" are removed because the object that was passed in doesn't have these annotations, so the syncer is correctly removing these. If we want to keep them we need to create a new object that has these as well, but I don't think this needs to be part of this testing logic. It worked before because the syncer was not updating metadata in this step and would require a second Sync call to remove that metadata

File: pkg/controllers/resources/persistentvolumeclaims/translate.go (hunk 3 above)
Comment (FabianKramm) [reply pkg/controllers/resources/persistentvolumeclaims/translate.go:107]: Oh yes true, good catch

File: pkg/controllers/resources/persistentvolumeclaims/syncer_test.go (hunk 1 above)
Comment (facchettos) [reply pkg/controllers/resources/persistentvolumeclaims/syncer_test.go:102]: makes sense, no need to keep it then I agree

File: pkg/controllers/resources/persistentvolumeclaims/translate.go (hunk 4 above)
Comment (FabianKramm) [reply pkg/controllers/resources/persistentvolumeclaims/translate.go:109]: We need that or otherwise these annotations would be always set, in the beginning both pObj and vObj don't have them, but this would result in an empty annotation on the virtual side

**Review by facchettos — APPROVED** [review]
